From db035ee8466fab56195125d3d132e59a8fde5987 Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Sun, 27 Mar 2022 12:22:25 -0700
Subject: [PATCH] ath12k: Fix TX completion ring mask for WBM2SW ring 4

Currently TX completions are not seen for packets transmitted
with return buffer programmed for 4th TX ring. This is causing
not freeing tx descriptors and eventually run of buffers. Fix
this by setting programming ring mask appropriately to match
wbm2sw ring id.

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp.c    | 10 +++++-----
 drivers/net/wireless/ath/ath12k/dp_tx.c |  1 +
 drivers/net/wireless/ath/ath12k/hw.c    |  2 +-
 3 files changed, 7 insertions(+), 6 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -79,6 +79,7 @@ static void ath12k_dp_tx_release_txbuf(s
 				       u8 pool_id)
 {
 	spin_lock_bh(&dp->tx_desc_lock[pool_id]);
+	tx_desc->skb = NULL;
 	list_move_tail(&tx_desc->list, &dp->tx_desc_free_list[pool_id]);
 	spin_unlock_bh(&dp->tx_desc_lock[pool_id]);
 }
--- a/drivers/net/wireless/ath/ath12k/hw.c
+++ b/drivers/net/wireless/ath/ath12k/hw.c
@@ -90,7 +90,10 @@ static const struct ath12k_hw_ops wcn785
 #define ATH12K_TX_RING_MASK_0 0x1
 #define ATH12K_TX_RING_MASK_1 0x2
 #define ATH12K_TX_RING_MASK_2 0x4
-#define ATH12K_TX_RING_MASK_3 0x8
+/* ATH12K_TX_RING_MASK_4 corresponds to the interrupt mask
+ * for the 4th TX Data ring which uses WBM2SW4RELEASE and
+ * hence uses the mask with 4th bit set
+ */
 #define ATH12K_TX_RING_MASK_4 0x10
 
 #define ATH12K_RX_RING_MASK_0 0x1
@@ -535,7 +538,7 @@ static const struct ath12k_hw_ring_mask
 		ATH12K_TX_RING_MASK_0,
 		ATH12K_TX_RING_MASK_1,
 		ATH12K_TX_RING_MASK_2,
-		ATH12K_TX_RING_MASK_3,
+		ATH12K_TX_RING_MASK_4,
 	},
 	.rx_mon_dest = {
 		0, 0, 0,
